common/issue-credential/src/utils.rs

```diff
@@ -0,0 +1,141 @@
+
+    if let Ok(recovered_amount) =
+        recover_credentials(&client, &recovery_storage, persistent_storage).await
+    {
+        if recovered_amount != 0 {
+            info!(
+                "Recovered credentials in the amount of {}",
+                recovered_amount
+            );
+            return;
+        }
+    };
+
+    let state = nym_bandwidth_controller::acquire::deposit(&client.nyxd, amount.clone())
+        .await
+        .expect("");
+
+    if nym_bandwidth_controller::acquire::get_credential(&state, &client, persistent_storage)
+        .await
+        .is_err()
+    {
+        warn!("Failed to obtain credential. Dumping recovery data.",);
+        match recovery_storage.insert_voucher(&state.voucher) {
+            Ok(file_path) => {
+                warn!("Dumped recovery data to {}. Try using recovery mode to convert it to a credential", file_path.to_str().unwrap());
+            }
+            Err(e) => {
+                error!("Could not dump recovery data to file system due to {:?}, the deposit will be lost!", e)
+            }
+        }
+        exit(1);
+    }
+
+    info!(
+        "Succeeded adding a credential with amount {}{}",
+        &amount.amount, &amount.denom,
+    );
+}
+
+pub async fn setup_recovery_storage(recovery_dir: PathBuf) -> RecoveryStorage {
+    RecoveryStorage::new(recovery_dir).expect("")
+}
+
+pub async fn setup_persistent_storage(client_home_directory: PathBuf) -> PersistentStorage {
+    let data_dir = client_home_directory.join(DEFAULT_DATA_DIR);
+    let paths = CommonClientPaths::new_default(data_dir);
+    let db_path = paths.credentials_database;
+
+    nym_credential_storage::initialise_persistent_storage(db_path).await
+}
+
+pub async fn block_until_coconut_is_available(
+    client: &Client<DirectSigningNyxdClient>,
+) -> Result<()> {
+    loop {
+        let epoch = client.nyxd.get_current_epoch().await?;
+        let current_timestamp_secs = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)?
+            .as_secs();
+        if epoch.state.is_final() {
+            if current_timestamp_secs + SAFETY_BUFFER_SECS >= epoch.finish_timestamp.seconds() {
+                info!("In the next {} minute(s), a transition will take place in the coconut system. Deposits should be halted in this time for safety reasons.", SAFETY_BUFFER_SECS / 60);
+                exit(0);
+            }
+
+            break;
+        } else {
+            // Use 1 additional second to not start the next iteration immediately and spam get_current_epoch queries
+            let secs_until_final = epoch
+                .final_timestamp_secs()
+                .saturating_sub(current_timestamp_secs)
+                + 1;
+            info!("Approximately {} seconds until coconut is available. Sleeping until then. You can safely kill the process at any moment.", secs_until_final);
+            std::thread::sleep(Duration::from_secs(secs_until_final));
```

Comment:
that's a rather subtle thing, but you don't want to be using `thread::sleep` in the async context. what you probably meant to use was `tokio::time::sleep` instead